roadmap/implementers-guide/src/node/disputes/votesdb.md

```diff
@@ -0,0 +1,137 @@
+
+## Remarks
+
+Votes must be persisted. Disputes might happen long after
+there was a vote, as such just keeping things in memory
+is thus insufficient, as every blip (OOM, bug, maintanance)
+could cause the node to forget the state.
+
+## ToDo
+
+* Missing Definition of Governance Mode.
+* Is a `CommittedCandidateReceipt` enough or should the whole `PoV` be included? There was talk about detachable validator signatures.
+* Currently [`fn mark_bad()`](https://github.com/paritytech/substrate/pull/6301/files#diff-8faeb5c685a8fdff428c5ec6d9102fd59e127ff69762d43045cd38e586db5559R60-R64) does not persist data.
+
+## IO
+
+Inputs:
+
+* `VotesDbMessage::DisputeVote`
+
+Outputs:
+
+* `DisputeParticipationMessage::Detection`
+* `DisputeParticipationMessage::Resolution`
+
+## Messages
+
+```rust
+enum VotesDbMessage {
+    /// Allow querying all `Hash`es queried by a particular validator.
+    QueryValidatorVotes {
+        /// Validator indentification.
+        session: SessionIndex,
+        validator: ValidatorIndex,
+        response: ResponseChannel<Vec<Hash>>,
+    },
+
+    /// Store a vote for a particular dispute
+    StoreVote{
+        /// Unique validator indentification
+        session: SessionIndex,
+        validator: ValidatorIndex,
+        /// Vote.
+        vote: Vote,
+        /// Attestation.
+        attestation: ValidityAttestation,
+    },
+}
+```
+
+```rust
+enum DisputeMessage {
+    /// A dispute is detected
+    Detection {
+        /// unique validator indentification
+        session: SessionIndex,
+        validator: ValidatorIndex,
+        /// The attestation.
+        attestation: ValidityAttestation,
+        /// response channel
+        response: ResponseChannel<Vec<Hash>>,
+    },
+    /// Concluded a dispute with the following resolution
+    Resolution {
+        /// Resolution of the block can either be,
+        /// the block being valid or not.
+        valid: bool,
+    }
+}
+```
+
+## Helper Structs
+
+```rust
+/// Snapshot of the current vote state and recorded votes
+struct Snapshot {
+    /// all entries must be unique
+    pro: Vec<(ValidatorIndex, SessionIndex, ValidityAttestation)>,
+    /// all entries must be unique
+    cons: Vec<(ValidatorIndex, SessionIndex, ValidityAttestation)>,
+}
+```
+
+## Session Change
+
+A sweep clean is to be done to remove stored attestions
+of all validators that are not slashable / have no funds
+bound anymore in order to save storage space.
+
+## Storage
+
+To fulfill the required query features.
+
+```raw
+(ValidatorIndex, Session) -> Vec<Hash>
+(ValidatorId) -> Vec<(ValidatorIndex, Session)>
+(Hash) -> (CommittedCandidateReceipt, PoV)
+```
+
+## Sequence
+
+### Incoming message
+
+1. Incoming request to store a dispute gossip message.
+1. Validate message.
+1. Check if the dispute already has a supermajority.
+    1. iff:
+        1. Craft and enqueue an unsigned transaction to reward/slash the validator of the proof, that is part of the incoming message, directly.
```

Comment:
I do believe this slashing should happen automatically on-chain. e.g. The goal of the node-side should just be to provide information to the various forks of the chain, which can act on it as necessary.

Reply:
It feels odd to split that up and defer it to the runtime when all the information to do so is already present.

Reply:
But `active_heads` is not enough I guess - that covers all the local forks, for forks that are not present locally (`remote` ones) how would those be best handled? (or is that the reason to do in the runtime?)

Reply:
> I do believe this slashing should happen automatically on-chain. e.g. The goal of the node-side should just be to provide information to the various forks of the chain, which can act on it as necessary.

In other words, the node sideonly crafts unsigned transactions for each active head and lets the runtime logic (block import) do whatever the runtime deems necessary.

Reply:
Slashing is something that's computed by the chain in our code.

> In other words, the node side only crafts unsigned transactions for each active head and lets the runtime logic (on block import? or another entry point?) do whatever the runtime deems necessary such as slashing.

Yes. And these transactions should be generated based on the data in that active head - if there are live disputes, or that chain has included blocks that have been disputed elsewhere. What the chain does with those can be coded in the runtime.

Reply:
Just to give a little bit of intuition on slashing. Here is my epistemology:
* Slashing has to happen somewhere
* Account balances and staking ledger are stored on-chain, so slashing must happen on-chain
* Slashing on-chain must be justified - the chain must be able to interpret data to determine that some offence has occurred
* Therefore the off-chain code should be concerned with producing justification as opposed to commands to slash
* In this case, justification takes the form of statements from many validators.